engine/ingame/controller.py

```diff
@@ -53,16 +53,28 @@ def on_mouse_release(
             self.model.activate()
 
     def on_update(self, delta_time: int) -> None:
+        """
+        Updates player speed based on keypress,
+        and prevents player going out-of-bounds.
+        """
         vx, vy = 0, 0
 
-        if self._up_key_pressed():
+        player_sprite = self.model.player_sprite
+        map_width = self.model.width * self.model.tile_width
+        map_height = self.model.height * self.model.tile_height
```

Comment:
As I mentioned in the ticket, I think this is more of a model concern than a controller concern. You can tell there's a code smell going on here because this class is accessing a bunch of children of the model and then manipulating them; the model is perfectly capable of handling this logic just as cleanly.

A problem with this approach too is it assumes that the only way to manipulate the player's speed is through the controller, however it will be possible for scripts to move the player too (imagine an explosion, spell, environment effect, or something like that).

A cleaner and more future-proof approach would be to simply project the movement as part of the model's `on_update` method:

```python
# in on_update:
new_x = self.player_sprite.center_x + self.player_sprite.change_x
new_y = self.player_sprite.center_y + self.player_sprite.change_y

if new_x < 0 or new_x > width:
  self.player_sprite.change_x = 0

# similarly for the Y axis
```

Reply:
> A problem with this approach too is it assumes that the only way to manipulate the player's speed is through the controller, however it will be possible for scripts to move the player too (imagine an explosion, spell, environment effect, or something like that).

That makes perfect sense. Thanks!